front/lib/api/data_sources.ts

```diff
@@ -211,3 +227,247 @@ export async function augmentDataSourceWithConnectorDetails(
+        )
+      );
+    }
+  }
+
+  // Enforce plan limits: DataSource document size.
+  if (
+    plan.limits.dataSources.documents.sizeMb != -1 &&
+    fullText.length > 1024 * 1024 * plan.limits.dataSources.documents.sizeMb
+  ) {
+    return new Err(
+      new DustError(
+        "data_source_quota_error",
+        `Data sources document upload size is limited to ` +
+          `${plan.limits.dataSources.documents.sizeMb}MB on your current plan. ` +
+          `You are attempting to upload ${fullText.length} bytes. ` +
+          `Contact support@dust.tt if you want to increase it.`
+      )
+    );
+  }
+
+  // Data source operations are performed with our credentials.
+  const credentials = dustManagedCredentials();
+
+  // Create document with the Dust internal API.
+  const upsertRes = await coreAPI.upsertDataSourceDocument({
+    projectId: dataSource.dustAPIProjectId,
+    dataSourceId: dataSource.dustAPIDataSourceId,
+    documentId: name,
+    tags: nonNullTags,
+    parents: parents || [],
+    sourceUrl,
+    timestamp: timestamp || null,
+    section: generatedSection,
+    credentials,
+    lightDocumentOutput: light_document_output === true,
+  });
+
+  if (upsertRes.isErr()) {
+    return new Err(
+      new DustError(
+        "core_api_error",
+        "There was an error upserting the document."
+      )
+    );
+  }
+
+  return new Ok(upsertRes.value);
+}
+
+export async function upsertTable({
+  tableId,
+  name,
+  description,
+  truncate,
+  csv,
+  tags,
+  parents,
+  timestamp,
+  async,
+  dataSource,
+  auth,
+}: {
+  tableId?: string | null;
+  name: string;
+  description: string;
+  truncate: boolean;
+  csv?: string | null;
+  tags?: string[] | null;
+  parents?: string[] | null;
+  timestamp?: number | null;
+  async: boolean;
+  dataSource: DataSourceType;
+  auth: Authenticator;
+}) {
+  const nonNullTableId = tableId ?? generateLegacyModelSId();
+  const tableParents: string[] = parents ?? [];
+
+  if (!tableParents.includes(nonNullTableId)) {
+    tableParents.push(nonNullTableId);
+  }
+
+  if (async) {
+    // Ensure the CSV is valid before enqueuing the upsert.
+    const csvRowsRes = csv ? await rowsFromCsv(csv) : null;
+    if (csvRowsRes?.isErr()) {
+      return csvRowsRes;
+    }
+
+    const enqueueRes = await enqueueUpsertTable({
+      upsertTable: {
+        workspaceId: auth.getNonNullableWorkspace().sId,
+        projectId: dataSource.dustAPIProjectId,
+        dataSourceName: dataSource.name,
+        tableId: nonNullTableId,
+        tableName: name,
+        tableDescription: description,
+        tableTimestamp: timestamp ?? null,
+        tableTags: tags ?? [],
+        tableParents,
+        csv: csv ?? null,
+        truncate,
+      },
+    });
+    if (enqueueRes.isErr()) {
+      return enqueueRes;
+    }
+
+    return new Ok(undefined);
+  }
+
+  const tableRes = await upsertTableFromCsv({
```

Comment:
@philipperolet is it expected that uploading a document on a static data source from the app does not go through. the upsert queue?

Reply:
for table we have an async param which allows to put it in the queue, but it is (was) set to false.